process/peer/process.go

```diff
@@ -0,0 +1,360 @@
+	for _, node := range in {
+		err := p.initializeNode(node)
+		if err != nil {
+			return err
+		}
+	}
+
+	_, err := p.peerAdapter.Commit()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// IsNodeValid calculates if a node that's present in the initial validator list
+//  contains all the required information in order to be able to participate in consensus
+func (p *validatorStatistics) IsNodeValid(node *sharding.InitialNode) bool {
+	if len(node.PubKey) == 0 {
+		return false
+	}
+	if len(node.Address) == 0 {
+		return false
+	}
+
+	return true
+}
+
+// UpdatePeerState takes the ca header, updates the peer state for all of the
+//  consensus members and returns the new root hash
+func (p *validatorStatistics) UpdatePeerState(header data.HeaderHandler) ([]byte, error) {
+	if header.GetNonce() == 0 {
+		return p.peerAdapter.RootHash()
+	}
+
+	consensusGroup, err := p.nodesCoordinator.ComputeValidatorsGroup(header.GetPrevRandSeed(), header.GetRound(), header.GetShardID())
+	if err != nil {
+		return nil, err
+	}
+
+	err = p.updateValidatorInfo(consensusGroup, header.GetShardID())
+	if err != nil {
+		return nil, err
+	}
+
+	previousHeader, err := p.getHeader(header.GetPrevHash())
+	if err != nil {
+		return nil, err
+	}
+
+	err = p.checkForMissedBlocks(header, previousHeader)
+	if err != nil {
+		return nil, err
+	}
+
+	err = p.updateShardDataPeerState(header)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.peerAdapter.RootHash()
+}
+
+// Commit commits the validator statistics trie and returns the root hash
+func (p *validatorStatistics) Commit() ([]byte, error) {
+	return p.peerAdapter.Commit()
+}
+
+// RootHash returns the root hash of the validator statistics trie
+func (p *validatorStatistics) RootHash() ([]byte, error) {
+	return p.peerAdapter.RootHash()
+}
+
+func (p *validatorStatistics) checkForMissedBlocks(currentHeader, previousHeader data.HeaderHandler) error {
+	if currentHeader.GetRound()-previousHeader.GetRound() <= 1 {
+		return nil
+	}
+
+	for i := previousHeader.GetRound() + 1; i < currentHeader.GetRound(); i++ {
+		consensusGroup, err := p.nodesCoordinator.ComputeValidatorsGroup(previousHeader.GetPrevRandSeed(), i, previousHeader.GetShardID())
+		if err != nil {
+			return err
+		}
+
+		leaderPeerAcc, err := p.getPeerAccount(consensusGroup[0].Address())
+		if err != nil {
+			return err
+		}
+
+		err = leaderPeerAcc.DecreaseLeaderSuccessRateWithJournal()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// RevertPeerState takes the current and previous headers and undos the peer state
+//  for all of the consensus members
+func (p *validatorStatistics) RevertPeerState(header data.HeaderHandler) error {
+	_ = p.peerAdapter.RecreateTrie(header.GetValidatorStatsRootHash())
+	return nil
+}
+
+// RevertPeerStateToSnapshot reverts the applied changes to the peerAdapter
+func (p *validatorStatistics) RevertPeerStateToSnapshot(snapshot int) error {
+	return p.peerAdapter.RevertToSnapshot(snapshot)
```

Comment:
I do not know if we should have an wrapper on top of an already existing interface. Metachain might hold the list of states and call revert / commit / roothash directly.